scalafix-core/shared/src/main/scala/scalafix/internal/util/TypeSyntax.scala

```diff
@@ -1,138 +1,451 @@
+    s.Type(s.Type.Tag.TYPE_REF, typeRef = Some(s.TypeRef(symbol = sym)))
+  }
 
-    /*
-     * Returns a scala.meta.Tree given a scala.meta.Symbol.
-     *
-     * Before: _root_.scala.Predef.Set#
-     * After: Type.Select(Term.Select(Term.Name("scala"), Term.Name("Predef")),
-     *                    Type.Name("Set"))
-     */
-    def symbolToTree(sym: Symbol.Global): (Patch, Tree) = {
-      var patch = Patch.empty
-      def loop[T: ClassTag](symbol: Symbol): T = {
-        val result = symbol match {
-          // base case, symbol `_root_.`  becomes `_root_` term
-          case Symbol.Global(Symbol.None, Signature.Term(name)) =>
-            Term.Name(name)
-          // symbol `A#B#`  becomes `A#B` type
-          case Symbol.Global(owner @ SymbolType(), Signature.Type(name)) =>
-            Type.Project(loop[Type](owner), Type.Name(name))
-          // symbol `A#B#`  becomes `A#B` type
-          case Symbol.Global(owner, Signature.Term(name)) =>
-            // TODO(olafur) implement lookup utility to query what symbol signature resolves to.
-            if (shortenNames && isStable(owner)) {
-              patch += ctx.addGlobalImport(symbol)
-              Term.Name(name)
-            } else {
-              Term.Select(loop[Term.Ref](owner), Term.Name(name))
-            }
-          // symbol `a.B#`  becomes `a.B` type
-          case Symbol.Global(owner, Signature.Type(name)) =>
-            if (shortenNames && isStable(owner)) {
-              patch += ctx.addGlobalImport(symbol)
-              Type.Name(name)
-            } else Type.Select(loop[Term.Ref](owner), Type.Name(name))
-
-          case Symbol.Global(_, Signature.TypeParameter(name)) =>
-            Type.Name(name)
+  def widen(tpe: s.Type): s.Type = {
+    tpe.tag match {
+      case t.SINGLETON_TYPE =>
+        import s.SingletonType.{Tag => x}
+        val singletonType = tpe.singletonType.get
+        singletonType.tag match {
+          case x.SYMBOL =>
+            index.denotation(Symbol(singletonType.symbol)).get.tpeInternal.get
+          case x.BOOLEAN => ref("scala.Boolean#")
+          case x.BYTE => ref("scala.Byte#")
+          case x.CHAR => ref("scala.Char#")
+          case x.DOUBLE => ref("scala.Double#")
+          case x.FLOAT => ref("scala.Float#")
+          case x.INT => ref("scala.Int#")
+          case x.LONG => ref("scala.Long#")
+          case x.NULL => ref("scala.Null#")
+          case x.SHORT => ref("scala.Short#")
+          case x.STRING => ref("java.lang.String#")
+          case x.UNIT => ref("scala.Unit#")
+          case x.SUPER => tpe
+          case x.THIS => tpe
+          case x.UNKNOWN_SINGLETON => tpe
+          case x.Unrecognized(_) => tpe
         }
+      // TODO: handle non-singleton widening.
+      case _ => tpe
+    }
+  }
+
+  /*
+   * Returns a scala.meta.Tree given a scala.meta.Symbol.
+   *
+   * NOTE: this method has become an utter mess and is in need of a clean reimplementation.
```

Comment:
Okay, not gonna review then.

Reply:
The implementation is unchanged from the previous TypeSyntax, the diff is mostly reformatting with minor changes.